RFS/src/test/java/com/rfs/integration/SnapshotStateTest.java

```diff
@@ -0,0 +1,248 @@
+                "  \"include_global_state\": false\n" +
+                "}";
+
+        final var createSnapshotRequest = new HttpPut(clusterUrl + "/_snapshot/test-repo/" + snapshotName + "?wait_for_completion=true");
+        createSnapshotRequest.setEntity(new StringEntity(snapshotJson));
+        createSnapshotRequest.setHeader("Content-Type", "application/json");
+
+        try (var response = httpClient.execute(createSnapshotRequest)) {
+            assertThat(response.getStatusLine().getStatusCode(), equalTo(200));
+        }
+
+        // Read snapshot file contents
+        for (File file : SNAPSHOT_DIR.listFiles()) {
+            System.out.println(file);
+        }
+        var snapshotFiles = Files.readAllLines(Paths.get(SNAPSHOT_DIR.getAbsolutePath(), "index-0"));
+        System.out.println("Snapshot Files: " + snapshotFiles);
+    }
+
+
+    @Test
+    public void SingleSnapshot_SingleDocument() throws Exception {
+        // Setup
+        // PSUEDO: Create an 1 index with 1 document
+        createDocument("my-index", "doc1", "{\"foo\":\"bar\"}");
+        // PSUEDO: Save snapshot1
+        final var snapshotName = "snapshot-1";
+        takeSnapshot(snapshotName, "my-index");
+        // PSUEDO: Start RFS worker reader, point to snapshot1
+        var repo = new FileSystemRepo(Path.of(SNAPSHOT_DIR.getAbsolutePath()));
+        var snapShotProvider = new SnapshotRepoProvider_ES_7_10(repo);
+        List<IndexMetadata.Data> indices = snapShotProvider.getIndicesInSnapshot(snapshotName)
+            .stream()
+            .map(index -> {
+                try {
+                    return new IndexMetadataFactory_ES_7_10().fromRepo(repo, snapShotProvider, snapshotName, index.getName());
+                } catch (final Exception e) {
+                    throw new RuntimeException(e);
+                }
+            })
+            .collect(Collectors.toList());
+        
+        var sourceFileBasePath = new File("./unpacked");
+
+        for (final IndexMetadata.Data index : indices) {
+            for (int shardId = 0; shardId < index.getNumberOfShards(); shardId++) {
+                var shardMetadata = new ShardMetadataFactory_ES_7_10().fromRepo(repo, snapShotProvider, snapshotName, index.getName(), shardId);
+                SnapshotShardUnpacker.unpack(repo, shardMetadata, Paths.get(sourceFileBasePath.getAbsolutePath()), Integer.MAX_VALUE);
+            }
+        }
+
+        var targetConnection = mock(ConnectionDetails.class);
+
+        for (final IndexMetadata.Data index : indices) {
+            for (int shardId = 0; shardId < index.getNumberOfShards(); shardId++) {
+                final var documents = new LuceneDocumentsReader().readDocuments(Paths.get(sourceFileBasePath.getAbsolutePath()), index.getName(), shardId);
+
+                final var finalShardId = shardId; // Define in local context for the lambda
+                DocumentReindexer.reindex(index.getName(), documents, targetConnection)
```

Comment:
@chelma I am going to need to start some refactoring to allow 'peaking' into this state.  I'll try to create a very specific PR targeting making the `reindex` take parameters for DI rather than creating its own client, etc..

Reply:
Sounds good, will be interested in what you find!